pincer/commands/chat_command_handler.py

```diff
@@ -0,0 +1,453 @@
+                        options=cmd.app.options
+                    )
+                )
+
+                continue
+
+            # All single-level commands are registered here.
+            ChatCommandHandler.built_register[
+                _hash_app_command(cmd.app, cmd.group, cmd.sub_group)
+            ] = cmd.app
+
+    @staticmethod
+    def get_local_registered_commands() -> ValuesView[AppCommand]:
+        return ChatCommandHandler.built_register.values()
+
+    async def __get_existing_commands(self):
+        """|coro|
+
+        Get AppCommand objects for all commands registered to discord.
+        """
+        try:
+            self._api_commands = await self.get_commands()
+
+        except ForbiddenError:
+            logging.error("Cannot retrieve slash commands, skipping...")
+            return
+
+    async def __remove_unused_commands(self):
+        """|coro|
+
+        Remove commands that are registered by discord but not in use
+        by the current client
+        """
+        local_registered_commands = self.get_local_registered_commands()
+
+        def should_be_removed(target: AppCommand) -> bool:
+            for reg_cmd in local_registered_commands:
+                # Commands have endpoints based on their `name` amd `guild_id`. Other
+                # parameters can be updated instead of deleting and re-registering the
+                # command.
+                if (
+                    target.name == reg_cmd.name
+                    and target.guild_id == reg_cmd.guild_id
+                ):
+                    return False
+            return True
+
+        # NOTE: Cannot be generator since it can't be consumed due to lines 743-745
+        to_remove = [*filter(should_be_removed, self._api_commands)]
+
+        await gather(
+            *map(
+                lambda cmd: self.remove_command(cmd),
+                to_remove,
+            )
+        )
+
+        self._api_commands = list(
+            filter(lambda cmd: cmd not in to_remove, self._api_commands)
+        )
+
+    async def __add_commands(self):
+        """|coro|
+
+        Add all new commands which have been registered by the decorator to Discord.
+
+        .. code-block::
+
+            Because commands have unique names within a type and scope, we treat POST
+            requests for new commands as upserts. That means making a new command with
+            an already-used name for your application will update the existing command.
+            `<https://discord.dev/interactions/application-commands#updating-and-deleting-a-command>`_
+
+        Therefore, we don't need to use a separate loop for updating and adding
+        commands.
+        """
+        local_registered_commands = self.get_local_registered_commands()
+
+        def should_be_updated_or_uploaded(target):
+            for command in self._api_commands:
+                if target == command:
+                    return False
+            return True
```

Comment:
```suggestion
        def should_be_updated_or_uploaded(target):
            return not any(target == command for command in self._api_commands)
```
```suggestion
        def should_be_updated_or_uploaded(target):
            return all(target != command for command in self._api_commands)
```

Reply:
I don't think changes this makes it more readable

Reply:
i refactored this to 
```
changed_commands = filter(
    lambda command: command not in self._api_commands, local_registered_commands
)
```
Can't believe that wasn't the first thing I thought of xD. sometimes the simple stuff evades you lol.